### PR TITLE
Autocomplete Support

### PR DESCRIPTION
**Pull request by vru26**
Previously the autocomplete functionality for input elements was limited and didn't behave as expected. This PR adds & updates the form autocomplete support for SingleLineTextBox which is basically the text input element. In order to test/preview these changes, the test app form test case has also been updated.

### CODE REVIEW

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 1]
```diff
@@ -71,7 +76,8 @@ export default class SingleLineTextBoxData extends InputTagData {
    * @returns {boolean} true if autocomplete is enabled (by default or explicitly), false otherwise
    */
   get autoComplete() {
-    return this._reactProps.autoComplete === undefined || this._reactProps.autoComplete === 'on';
+    if (this._reactProps.autoComplete !== undefined) return this._reactProps.autoComplete;
```
Comment (helenxuyang) [inline src/RoleObjects/SingleLineTextBoxData.js:79]: I'm wondering it's confusing if the getter can return either a bool or string. Maybe if we change the setter to only take strings, we should change the getter to only take strings as well.

Also, the undefined case is kind of confusing, but I believe if autocomplete is undefined, it's enabled by default (according to https://www.w3schools.com/tags/att_form_autocomplete.asp and https://www.w3schools.com/tags/att_input_autocomplete.asp) so we should technically return 'on'..? This is kind of confusing and not well-documented though, so I'm not sure.

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 2]
```diff
@@ -62,7 +62,12 @@ export default class SingleLineTextBoxData extends InputTagData {
    * @param {boolean} enable - true if autocomplete should be enabled, false otherwise
```
Comment (helenxuyang) [inline src/RoleObjects/SingleLineTextBoxData.js:62]: Just a reminder that we'll want to update the doc comments for both the setter and getter to match our changes

File: test-app/src/widgets/AppWindow.js [hunk 3]
```diff
@@ -809,22 +909,28 @@ export default class AppWindow extends createjs.Container {
...
 
     const submitBtnData = {
-      type: 'button',
+      type: 'submit',
       value: 'SUBMIT',
       name: 'SUBMIT',
       enabled: true,
       autoFocus: false,
-      form: 'form_1',
+      // form: 'form_1',
```
Comment (helenxuyang) [inline test-app/src/widgets/AppWindow.js:924]: Just a question cause I'm trying to understand how the button and form work, why is this commented out?

File: test-app/src/widgets/AppWindow.js [hunk 4]
```diff
@@ -852,6 +958,7 @@ export default class AppWindow extends createjs.Container {
       height: 60,
       width: 250,
     };
+    let reset;
```
Comment (helenxuyang) [inline test-app/src/widgets/AppWindow.js:961]: Just trying to understand, why was this changed from using const to using let?

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 5]
```diff
@@ -62,7 +62,12 @@ export default class SingleLineTextBoxData extends InputTagData {
...
    */
   set autoComplete(enable) {
-    this._reactProps.autoComplete = enable ? 'on' : 'off';
+    const validValues = ['off', 'on', 'name', 'email', 'username', 'new-password', 'current-password', 'one-time-code', 'organization-title', 'organization', 'street-address', 'address-line1', 'address-line2', 'address-line3', 'address-level4', 'address-level3', 'address-level2', 'address-level1', 'country', 'country-name', 'postal-code', 'cc-name', 'cc-given-name', 'cc-additional-name', 'cc-family-name', 'cc-number', 'cc-exp', 'cc-exp-month', 'cc-exp-year', 'cc-csc', 'cc-type', 'transaction-currency', 'transaction-amount', 'language', 'bday', 'bday-day', 'bday-month', 'bday-year', 'sex', 'tel', 'tel-extension', 'impp', 'url', 'photo'];
+    if (typeof (enable) === 'boolean') {
```
[the remark this answers is not in the document]
Comment (mmoss) [reply src/RoleObjects/SingleLineTextBoxData.js:66]: This keeps the change backwards compatible API-wise.

[the remark this answers is not in the document]
Comment (helenxuyang) [reply src/RoleObjects/SingleLineTextBoxData.js:66]: ahh gotcha, I think that clears out most of my comments then!

File: test-app/src/widgets/AppWindow.js (hunk 4 above)
Comment (vru26) [reply test-app/src/widgets/AppWindow.js:961]: This change was done to resolve a listing error. Since `reset` is been used on lines 973 & 1005 before it's been actually created/initialized hence a listing error was thrown.

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 6]
```diff
@@ -62,7 +62,12 @@ export default class SingleLineTextBoxData extends InputTagData {
...
+      this._reactProps.autoComplete = enable ? 'on' : 'off';
+    } else {
+      this._reactProps.autoComplete = validValues.includes(enable) ? enable : 'off';
```
[the remark this answers is not in the document]
Comment (vru26) [reply src/RoleObjects/SingleLineTextBoxData.js:69]: What type of exception should be thrown here? I am unsure about that hence I though setting it to off will probably be the safest way.

File: src/RoleObjects/SingleLineTextBoxData.js (hunk 2 above)
Comment (vru26) [reply src/RoleObjects/SingleLineTextBoxData.js:62]: Yeah I can update those once the backward compatibility stuff is sorted out.

**Review by mmoss — CHANGES_REQUESTED** [review]

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 7]
```diff
@@ -62,7 +62,12 @@ export default class SingleLineTextBoxData extends InputTagData {
    * @param {boolean} enable - true if autocomplete should be enabled, false otherwise
    */
   set autoComplete(enable) {
```
[the remark this answers is not in the document]
Comment (vru26) [reply src/RoleObjects/SingleLineTextBoxData.js:64]: Since I think we are trying to keep it backward compatible as well, If I change the function signature, won't this create backward compatibility issue?

File: test-app/src/widgets/AppWindow.js (hunk 3 above)
Comment (vru26) [reply test-app/src/widgets/AppWindow.js:924]: I did comment that out initially because I thought it might be creating an issue with the form submission since the `form` attribute takes the form id and form id isn't `form_1`. Form id is actually `acc_167`.

File: src/RoleObjects/SingleLineTextBoxData.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (mmoss) [reply src/RoleObjects/SingleLineTextBoxData.js:64]: Changing the signature (specifically the variable names), when it's not a function that takes an object with expected keys, doesn't change the API. The function signature doesn't _technically_ change.

[the remark this answers is not in the document]
Comment (helenxuyang) [reply src/RoleObjects/SingleLineTextBoxData.js:64]: We should also mark that it's now boolean or string, right?

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 8]
```diff
@@ -59,19 +59,27 @@ export default class SingleLineTextBoxData extends InputTagData {
...
   /**
    * Retrieves whether autocomplete is enabled
    * @access public
-   * @returns {boolean} true if autocomplete is enabled (by default or explicitly), false otherwise
+   * @returns {string | boolean} attribute value if autocomplete is enabled (by default or explicitly), false otherwise
    */
   get autoComplete() {
-    return this._reactProps.autoComplete === undefined || this._reactProps.autoComplete === 'on';
+    if (this._reactProps.autoComplete !== undefined) return this._reactProps.autoComplete;
+    return false;
```
Comment (helenxuyang) [inline src/RoleObjects/SingleLineTextBoxData.js:82]: I think my only remaining question is about whether this should return true if autoComplete is enabled by default, double checking with @mmoss since it's a bit of a weird situation where it's undefined which isn't exactly true?

Comment (mmoss) [reply src/RoleObjects/SingleLineTextBoxData.js:82]: I think this makes sense, since `undefined` does indeed mean that the autocomplete attribute hasn't been set, and the expected behavior is for autocomplete to be enabled.

**Review by helenxuyang — DISMISSED** [review]
Two minor comments about doc comments, otherwise I think this is good to go!! Thanks Vrushabh for the huge amount of work you put into this PR!

File: src/RoleObjects/SingleLineTextBoxData.js [hunk 9]
```diff
@@ -59,19 +59,27 @@ export default class SingleLineTextBoxData extends InputTagData {
   /**
    * Sets whether the form should have autocomplete enabled
    * @access public
-   * @param {boolean} enable - true if autocomplete should be enabled, false otherwise
-   */
-  set autoComplete(enable) {
-    this._reactProps.autoComplete = enable ? 'on' : 'off';
+   * @param {string | boolean} value - valid autocomplete attribute value to enabled, false or off otherwise
```
[the remark this answers is not in the document]
Comment (vru26) [reply src/RoleObjects/SingleLineTextBoxData.js:62]: Yeah. I will update that typo.

**Review by helenxuyang — APPROVED** [review]
LGTM, thank you Vrushabh for all your hard work on this PR!!
